test/frame.spec.js

```diff
@@ -162,6 +162,17 @@ module.exports.addTests = function({testRunner, expect}) {
       expect(detachedFrames.length).toBe(4);
       expect(navigatedFrames.length).toBe(1);
     });
+    it('should report frame from-inside shadow DOM', async({page, server}) => {
+      await page.goto(server.PREFIX + '/shadow.html');
+      await page.evaluate(async url => {
+        const frame = document.createElement('iframe');
+        frame.src = url;
+        document.body.shadowRoot.appendChild(frame);
+        await new Promise(x => frame.onload = x);
+      }, server.EMPTY_PAGE);
+      expect(page.frames().length).toBe(2);
```

Comment:
I think this might be flaky if you don’t still wait for the frameattached event

Reply:
shouldn't be - we await the "load" event on the iframe which happens way after the frameattached protocol event.